Update contribution guide and add issue/PR templates (#64)
PretendoNetwork/Pretendo: .github/CONTRIBUTING.md

Given that our project's main goal is to emulate Nintendo Network, it is also highly recommended that you familiarize yourself with how these services operate. Information on these topics, and our libraries, can be found both on our [wiki](https://nintendo-wiki.pretendo.network/docs/) and our (in progress) [developer documentation](https://developer.pretendo.network/home).

> [!NOTE]
> Developer documentation is highly work in progress. Do not expect to find everything there at this time. Please refer to the wiki first and foremost.

# Questions and Contact
If you have questions about a specific topic or wish to contact the core development team, you have 3 main options:

1. GitHub issues (recommended for specific questions)
2. The forum (recommended for broader/more general questions and discussion)
3. The Discord server (not recommended)

If you have a specific question regarding a specific service, tool, etc. then it is preferred if you make an issue on the relevant repository using the `Question` issue template. An example would be asking a question regarding the implementation of a specific RMC method on a specific game server.

> If you have a more generalized question then it is preferred if you create a thread on the [Technical Discussion forum](https://forum.pretendo.network/c/technical-discussion/5). An example would be asking about a topic which spans multiple services, clarifying questions about infrastructure, contacting the team regarding good first steps/issues, etc. The [Technical Discussion forum](https://forum.pretendo.network/c/technical-discussion) is also a place for contributors, both current and potential, to have general discussions with each other regarding Pretendo Network. This is arguably the best place to contact the core development team.

Reviewer: ```suggestion
If you have a more generalized question then it is preferred if you create a thread on the [Technical Discussion forum](https://forum.pretendo.network/c/technical-discussion). An example would be asking about a topic which spans multiple services, clarifying questions about infrastructure, contacting the team regarding good first steps/issues, etc. The [Technical Discussion forum](https://forum.pretendo.network/c/technical-discussion) is also a place for contributors, both current and potential, to have general discussions with each other regarding Pretendo Network. This is arguably the best place to contact the core development team.
```

If a pull request requires any database migrations, describe them in detail and leave any migration queries inside of a code block within a `<details>` tag. This should happen either at the very beginning of the pull request message, or at the very end, but not somewhere in between. Doing so makes it clear at a glance that there are migrations required and makes it easy to find the related queries.

We ask that you have patience with us as we review your pull request. Pretendo Network only has a single full time developer, all other work is done by volunteers on their own time. Due to the sheer number of issues and pull requests, alongside our other general work and research, it may take us some time to fully review and decide on whether or not to merge your changes.

# Tests
We do not require 100% code coverage in any tests. We do not currently have strict rules regarding tests, however we may ask that tests be provided for large or complex changes.

# Licensing

> Unless otherwise specified all code is licensed under [GNU AGPLv3](https://choosealicense.com/licenses/agpl-3.0), including that of outside contributions. This license allows users many freedoms to use our code in their own applications, including private and commercial use, while ensuring that all derivatives remain under this same license and keeps the source available, even when used over a network. A repositories license may not be changed by outside contributions unless permitted by the core development team, with good reason, and the consent of all relevant contributors has been given.

Reviewer 2: The last sentence's phrasing seems awkward.

```suggestion
Unless otherwise specified all code is licensed under [GNU AGPLv3](https://choosealicense.com/licenses/agpl-3.0), including that of outside contributions. This license allows users many freedoms to use our code in their own applications, including private and commercial use, while ensuring that all derivatives remain under this same license and keeps the source available, even when used over a network. A repository's license may not be changed by outside contributors unless that change is done with good reason, has been approved by the core development team, and is done with the consent of all relevant contributors.
```

Author: lgtm

- [Code Changes](#code-changes)
	- [Consistency](#consistency)
	- [Quality and Substance](#quality-and-substance)
- [Commits](#commits)
	- [Scale and Scope](#scale-and-scope)
	- [Messages](#messages)
- [Pull Requests](#pull-requests)

# Prerequisites
Before diving into the contribution process, we recommend familiarizing yourself with our project's existing repositories. By studying our existing codebases, issues, and pull requests you will get a better sense of direction of how we operate. Additionally, please review our [Code of Conduct](CODE_OF_CONDUCT.md) to ensure a respectful and inclusive environment for all participants.

Given that our project's main goal is to emulate Nintendo Network, it is also highly recommended that you familiarize yourself with how these services operate. Information on these topics, and our libraries, can be found both on our [wiki](https://nintendo-wiki.pretendo.network/docs/) and our (in progress) [developer documentation](https://developer.pretendo.network/home).

> [!NOTE]
> Developer documentation is highly work in progress. Do not expect to find everything there at this time. Please refer to the wiki first and foremost.

> Finally, we recommend joining our [Discord server](https://invite.gg/pretendo) to see any roadmap plans or ask specific questions from our developers.

Reviewer 3: Joining the discord is going to expose you to #pretendo-general, and it's really not a good way to reach out to us due to the flooding issues.

Author: I'm open to alternatives if you have them. I couldn't think of any other real way to do this. So far Discord has been the best way to contact me, at least, and some devs have (devs wanting to use our servers for their projects or our tools). Genuinely asking, do you have a better idea?

Reviewer 2: It could be made a bit clearer that it is completely okay to ping a developer for genuine technical questions about the Pretendo servers, although it would be good to keep that in support threads instead of general. I didn't understand that until I joined the supporter chat and saw how willing you all were to answer questions because it's hard to see any of that with how fast general chat moves to other topics (even before the NN closure joins).

Author: That sounds fair 👍 if you have any specific suggestions for how to word this feel free to make a commit suggestion here

Reviewer 3: I have had some luck redirecting technical questions into #support and having them still ping a dev, since it's at least easier to hold a conversation in its own thread. It's still not perfect because it won't get seen there unless OP finds the right person to ping, but it's better..

> Finally, we recommend joining our [Discord server](https://invite.gg/pretendo) to see any roadmap plans or ask specific questions from our developers.

Reviewer 2: ```suggestion
Finally, we recommend joining our [Discord server](https://invite.gg/pretendo) to see any roadmap plans or ask specific questions from our developers. To ask a technical question, please create a thread in the [#support channel](https://discord.com/channels/408718485913468928/1067162922989662360) and @mention a relevant developer who has recently committed to the repository in question. If you don't know who to mention or if it is a general question about multiple servers, mention @jonbarrow.
```

Reviewer 2: Might need to clarify which questions are more appropriate for a GitHub issue using the question template vs a discord question. Or just the user's choice.

Author: > could this be accepted then updated if there is a change to the public channels/server?

The idea itself is pretty solid but we haven't settled on the best way to go about this yet, which is why this hasn't been accepted. The support channel is *intended* for people with general service issues, not necessarily technical discussions. I'm not sure that trying to reuse that channel specifically would do much good, as it would just start to clog the channel up a bit and wouldn't really solve the issue of getting developer attention (this was mentioned on Discord yesterday, but we tend to opt for a "community-first" style of support right now due to lack of staff. The community helps each other in those threads, and we only jump in when need be. Many of us don't even watch those threads ourselves, so any technical questions would have the same level of lacking visibility)

Honestly one of the biggest criticisms we get is people feeling like we rely on our Discord server for far too much (people dislike that it's one of the main ways to get support, since not everyone uses Discord). I've been thinking for a while now about just moving a lot of this out of Discord and into something like an online forum that we'd run. Moving stuff to our own forum on our website would solve several issues:

- Support outside of Discord
- Appeals for things like bans
- More organized player reports
- Asking technical questions

We can easily split these topics into their own categories on a forum to keep things organized, which would help with visibility a lot I think. I haven't brought this idea up with the rest of the team, but now is probably a good time

@DaniElectra @shutterbug2000 @SuperMarioDaBom @hauntii @gitlimes @ashquarky @InternalLoss @TraceEntertains @MatthewL246

Reviewer 2: > The support channel is intended for people with general service issues, not necessarily technical discussions.

It probably does make sense to keep these things separate. Let the community continue to help people who need more basic help with getting things set up or fixing errors, and keep the technical discussion in a place with a lower volume of posts so developers can keep track of them without needing to ping people.

> moving a lot of this out of Discord and into something like an online forum that we'd run.

A forum sounds like a good idea to me. To be completely honest, I have no experience in contributing to online forums, but I really appreciate how they often provide great information about niche topics because they're public and indexed.

I'm conflicted about how organizing should work: I think that there are a lot of people on Discord who would be unwilling to sign up for an account on a forum, so it makes sense to keep the support channel available for them. (Not sure if any forum software has a "Log in with Discord" feature?) But the additional friction for joining a forum might help keep technical discussions higher quality? However, that also partially defeats the purpose of a forum, as people who don't use Discord will still be left in the dark on potentially useful information.

Author: > unwilling to sign up for an account on a forum

There's no need, we have that solved already; the PNID. This isn't my first time doing this, actually. Around 10 years ago I was the lead of a massive vanilla MMORPG Minecraft server, and we had our own custom forum that I wrote which used users Minecraft accounts as their forum accounts

Reviewer 2: For some reason I didn't think about that, makes total sense!

# Making Issues
Before contributing any code, you should understand how we use issues and how to make them in a way which ensures you will get the response you want.

> Issues are used by all repositories for most interactions. Issues are used for reporting actual issues with the codebase, as well as making feature requests and asking general questions. When making an issue, please use one of the provided issue templates. Using the provided templates keeps things consistent, as well as makes it easier for any of our 3rd party tools to interact with issues. If a template does not exist for what you wish to do, create a feature request for one. Not using an issue template may result in your issue being closed without completion.

Reviewer 3: > If a template does not exist for what you wish to do, create a feature request for one. Not using an issue template may result in your issue being closed without completion.

This feels like an unnecessary barrier to contribution, implying that they have to wait for a feature request to be implemented before even reporting their issue.

Author: There are issue templates for all the common use cases (even going so far as to split bug reports into bugs and regressions). The logic here is that if you need to use issues for something not covered by the existing templates then it shows we may need to add one

The point of the templates is to make it so that issues are structured the same across all issues, which should make reading them faster as we'd know where the relevant information is already

Do you have a better suggestion? (Genuine question)

Also to be clear the "it may be closed without completion" part was mostly aimed at people making issues that have templates available and just don't use them (like filing a bug report without using the template)

Reviewer 3: Maybe ask to please *also* file a feature request? So they can still file their thing immediately.

Reviewer 4: Take a small PR to implement a small fix - like, a typo in a print statement, do we *really* need an issue for that?

Do we *need* an issue with no comment or discussion for a PR that will definitely be merged? I feel like its unnecessary - issues IMO should be for 'these need to be implemented in the future but I'm not going to do it now'

Requiring changes to be approved and having substance is essential to not wasting the time of both contributors (who may spend time making changes we are not interested in) and our developers (who will have to spend time reviewing changes which ultimately get rejected).

# Commits
Besides the changes themselves, commits are the most important part of contributions. There are 2 major things to keep in mind for commits:

1. [Scale/Scope](#scale-and-scope)
2. [Messages](#messages)

## Scale and Scope
The scale and scope of a commit should be reasonable. Do not commit for every line when making multiple changes, for example. However you should not include many unrelated changes in a single commit. By limiting the scope of the commit we can ensure that if any regressions or new bugs are introduced we can easily revert those changes without the need for major refactors or reimplementations. Limiting the scale of commits also makes review of the changes easier and faster.

## Messages

> Commit messages should adequately explain the changes in the commit. Messages like "Updated file.md" and "spelling error" should not be used. Nonsense messages such as "oops" or "fixed" are especially not allowed. Commit messages should, at minimum, be in the format `type: message` where `type` represents the type or scope of the changes (`chore`, `docs`, `bug-fix`, etc.) and `message` is the actual changes. Unless the word is from the codebase and starts with a capital letter (such as an exported Go struct), the `message` should be lowercase. We also recommend using both "subject" and "body" commits. This can be achieved through the git CLI by using multiple `-m`/`--message` flags. For example `git commit -m "short subject" -m "longer description of the changes"`.

Reviewer: > the `message` should be lowercase

Unsure about having this as a requirement tbh. I personally prefer the subject being capitalized since it makes it stand out more from the description imo

The [Linux kernel](https://lore.kernel.org/asahi/) allows both lowercase and capitalized, can we also do this?

Author: This section was mostly inspired by https://conventionalcommits.org/, which doesn't say messages should be upper or lower case. In fact it says either acceptable, it just says to be consistent. Given that currently our existing messages seem to always be lowercase that's just the ruling I went with here, to keep things consistent. If that was an incorrect judgement I'm open to changing it. The casing itself is not that important to me, just the consistency part

The following are examples of good commit messages:

- `chore: renamed nnid service to nnas`
- `bug-fix: fixed hang in MutexMap.Has`

Please refer to [Conventional Commits](https://conventionalcommits.org/) for a detailed guide on how to structure commit messages. Writing good, detailed, commit messages helps ensure that we can refer back to the git history and quickly find where specific changes occurred in the event that they need further review, reverting, etc.

# Pull Requests
As stated in [Making Issues](#making-issues), before making a pull request there must be an open, approved, issue for the changes being made. If you can not find an approved issue for the changes you wish to make, please make one before continuing.

Before making a pull request ensure you have tested all changes and that no regressions have been introduced.

> Pull requests should never be made against the default (`main`/`master`) branch of a repository. The default branch contains the most recent, stable, version of the codebase. All work on the codebase should take place in other branches. Unless otherwise specified, your target branch should typically be the `dev` branch. You may target other feature branches, however, if need be. If a `dev` branch does not exist for the repository you are working on, please submit a feature request for one to be added before continuing.

Reviewer 3: I don't have `dev` branches on Inkay at least, we all commit to `main` and make releases for the stable build. Should I learn a new workflow? Can this be enforced via branch protection/PR templates somehow?

Author: Making a `dev` branch on all the repos is something we've been talking about for quite some time now (like a year or so). So this was going to be brought up either way

Your flow wouldn't really change all that much. Just working in `dev` rather than `main` and when a release is ready to be made merging and making the release. It's almost identical just with an extra step

> Can this be enforced via branch protection/PR templates somehow?

I may be stupid, wdym?

Reviewer 3: Like, forcing PRs to be against dev rather than main.

Author: Yes this can be done via branch protection. There's already a handful of repositories in the org which protect the default branch from direct commits and require PRs, so that would be added as well

The point about targeting the dev branches was there not only to tell people not to even try and PR against the default branch, but to also not base their changes off of it (which isn't something we can directly control. The most we can do is tell them not to in a document like this, but we can't physically stop someone from making a branch to make changes based off the default branch locally)

> Like everything else, pull requests should be as detailed as possible. Your title should adequately summarize the changes being made, and the body of the pull request should fully explain your changes. We ask that, if applicable, the rationale behind your changes also be noted. For example rather than simply "Changed from `for...of` to `forEach`", if the change was made for a performance reason you should say "Changed from `for...of` to `forEach` due to `forEach` being X times faster in this case" and provide some benchmarks. Adding images, videos, etc. is also welcomed in order to illustrate changes. If the changes being made are directly tied to some form of visual (such as a change to the website, a tools GUI, etc.) then images or videos is ***REQUIRED***. If none are provided, then we may delay review until they are given. Providing visual examples of these changes allow us to quickly assess whether or not we wish to proceed with the changes being made.

Reviewer 3: It's not obvious how to benchmark most of our stuff - e.g. NEX servers under high load - so is this actually a case for us to write up test harnesses? Is that a good use of our time right now?

Reviewer 3: Right, so providing benchmarks for performance changes isn't a rule, that was an example of providing details for evaluation.

Author: It was an example of how to explain the reasoning behind the changes. Nothing more and nothing less. I'm not sure what the issue is with this?

The point was to explain why changes were made if applicable. The example being changes made for performance reasons. If someone makes a change for performance reasons, it stands to reason that in that example it's because of something they could be measured (otherwise how would the performance issue even be found and known to be fixed). Thus in the example you could provide a benchmark for the changes

Again, I'm not understanding the issue here? It was literally just an example situation that I made up to illustrate a point

Reviewer 3: Yup, no issue. Was just making sure I understood what you said correctly.

Author: My apologies then! I thought you still had some sort of concern about it, I misunderstood

The scale and scope of a commit should be reasonable. Do not commit for every line when making multiple changes, for example. However you should not include many unrelated changes in a single commit. By limiting the scope of the commit we can ensure that if any regressions or new bugs are introduced we can easily revert those changes without the need for major refactors or reimplementations. Limiting the scale of commits also makes review of the changes easier and faster.

## Messages

> Commit messages should adequately explain the changes in the commit. Messages like "Updated file.md" and "spelling error" should not be used. Nonsense messages such as "oops" or "fixed" are especially not allowed. Commit messages should, at minimum, be in the format `type: message` where `type` represents the type or scope of the changes (`feat`, `chore`, `docs`, `bug-fix`, etc.) and `message` is the actual changes. Unless the word is from the codebase and starts with a capital letter (such as an exported Go struct), the `message` should be lowercase. We also recommend using both "subject" and "body" commits. This can be achieved through the git CLI by using multiple `-m`/`--message` flags. For example `git commit -m "short subject" -m "longer description of the changes"`.

Reviewer 2: I noticed https://www.conventionalcommits.org/en/v1.0.0/ uses `fix`, but you recommend `bug-fix` here.

Author: That's just an unfortunate example I chose. The guidelines allow other commit types besides `feat` and `fix` so I was trying to give examples of those (and our allowed formatting using dashes), and just accidentally conflicted with an existing type
